launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,457 @@
+    try:
+        fspath = inspect.getsourcefile(obj)
+    except TypeError:
+        return 'location information of the object not found'
+    try:
+        lines, lineno = inspect.getsourcelines(obj)
+    except IOError:
+        return f'file {fspath}: source code not available'
+    error_msg = f'file {fspath}, line {lineno}'
+    for line in lines:
+        line = line.rstrip()
+        error_msg += f'\n  {line}'
+        if line.lstrip().startswith('def'):
+            break
+    return error_msg
+
+
+@pytest.mark.trylast
+def pytest_collection_modifyitems(session, config, items):
+    """Reorder tests, so shutdown tests happen after the corresponding fixture teardown."""
+
+    def cmp(left, right):
+        leftm = left.get_closest_marker('launch_testing')
+        rightm = right.get_closest_marker('launch_testing')
+        if None in (leftm, rightm):
+            return 0
+        fixture = leftm.kwargs['fixture']
+        if fixture is not rightm.kwargs['fixture']:
+            return 0
+        if fixture._pytestfixturefunction.scope == 'function':
+            return 0
+        left_is_shutdown = int(left._launch_testing_is_shutdown)
+        right_is_shutdown = int(right._launch_testing_is_shutdown)
+        return left_is_shutdown - right_is_shutdown
+
+    # python sort is guaranteed to be stable
+    items.sort(key=functools.cmp_to_key(cmp))
+    # This approach doesn't work, because we cannot guarantee when fixture
+    # finalizers are going to be run.
+    # remove the launch_testing fixture from shutdown tests
+    # so launch service shutdown happens right before... by magic!!!
+    for item in items:
+        marker = item.get_closest_marker('launch_testing')
+        if marker is not None and item._launch_testing_is_shutdown:
+            item.fixturenames.remove(marker.kwargs['fixture'].__name__)
+
+
+@pytest.hookimpl(hookwrapper=True, tryfirst=True)
+def pytest_pyfunc_call(pyfuncitem):
+    """Run launch_testing test coroutines and functions in an event loop."""
+    marker = pyfuncitem.get_closest_marker('launch_testing')
+    if marker is not None:
+        args = {}
+        func = pyfuncitem.obj
+        spec = inspect.getfullargspec(func)
+        shutdown_test = pyfuncitem._launch_testing_is_shutdown
+        extra_args = {}
+        fixture = marker.kwargs['fixture']
+        scope = fixture._pytestfixturefunction.scope
+        if not shutdown_test:
+            fixturename = marker.kwargs['fixture'].__name__
+            ld_extra_args_pair = pyfuncitem.funcargs[fixturename]
+            if isinstance(ld_extra_args_pair, tuple) and len(ld_extra_args_pair) == 2:
+                extra_args = ld_extra_args_pair[1]
+        event_loop = pyfuncitem.funcargs['event_loop']
+        ls = pyfuncitem.funcargs['launch_service']
+        on_shutdown = functools.partial(
+            finalize_launch_service, ls, eprefix=f'When running test {func.__name__}')
+        before_test = on_shutdown if shutdown_test else None
```

Comment:
> meta: immediate fixture shutdown for pure post-shutdown tests? There may not even be enough time for anything meaningful to happen

True

> I'm inclined to think that such tests go hand-in-hand with self-terminating fixtures.

What do you mean?

Reply:
Oh, I never replied. I mean that a pure-shutdown test would only make sense with a fixture that terminates itself (by going idle with `shutdown_on_idle=True`, by emitting a `Shutdown()` event) after some time. If the fixture doesn't terminate itself, what will? when?

Reply:
> Oh, I never replied. I mean that a pure-shutdown test would only make sense with a fixture that terminates itself (by going idle with shutdown_on_idle=True, by emitting a Shutdown() event) after some time. If the fixture doesn't terminate itself, what will? when?

The fixture teardown step or the shutdown tests, both are currently shutting down the launc service.
The only issue is that nothing will setup the launch service, because pure shutdown tests don't depend on the launch description fixture.

Reply:
> The only issue is that nothing will setup the launch service, because pure shutdown tests don't depend on the launch description fixture.

this is now fixed, in a pure shutdown test we're now launching the fixture and waiting to shutdown to run it